src/server/editorServices.ts

```diff
@@ -1217,13 +1217,35 @@ namespace ts.server {
             }
             else {
                 const project = this.createProject(configFilename, projectOptions);
+                let programSize = 0;
+
+                // As the project openning might not be complete if there are too many files,
+                // therefore to surface the diagnostics we need to make sure the given client file is opened.
+                if (clientFileName) {
+                    const currentClientFileInfo = this.openFile(clientFileName, /*openedByClient*/ true);
+                    project.addRoot(currentClientFileInfo);
+                    programSize += currentClientFileInfo.content.length;
+                }
+
                 for (const rootFilename of projectOptions.files) {
-                    if (this.host.fileExists(rootFilename)) {
-                        const info = this.openFile(rootFilename, /*openedByClient*/ clientFileName == rootFilename);
-                        project.addRoot(info);
+                    if (rootFilename === clientFileName) {
+                        continue;
+                    }
+
+                    if (programSize <= maxProgramSize) {

```

Comment:
Because tsserver [opens up files](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L1222) before [calling the `createProgram` via `compilerService`](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L1229). And its [`openFile`](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L991) function does some pretty complex logic for file version management. In the `createProgram`, the compiler calls `findSourceFile` function, which will load the file if it found the file wasn't loaded. Therefore I need to set the limit on both sides. I didn't find an more optimized solution to unify them yet.